driver-core/src/main/com/mongodb/internal/connection/OperationContext.java

```diff
@@ -15,20 +15,103 @@
+     */
+    public ServerDeprioritization getServerDeprioritization() {
+        return serverDeprioritization;
+    }
+
+    public static final class ServerDeprioritization {
+        @Nullable
+        private ServerAddress candidate;
+        private final Set<ServerAddress> deprioritized;
+        private final DeprioritizingSelector selector;
+
+        private ServerDeprioritization() {
+            candidate = null;
+            deprioritized = new HashSet<>();
+            selector = new DeprioritizingSelector();
+        }
+
+        /**
+         * The returned {@link ServerSelector} tries to {@linkplain ServerSelector#select(ClusterDescription) select}
+         * only the {@link ServerDescription}s that do not have deprioritized {@link ServerAddress}es.
+         * If no such {@link ServerDescription} can be selected, then it selects {@link ClusterDescription#getServerDescriptions()}.
+         */
+        ServerSelector getServerSelector() {
+            return selector;
+        }
+
+        void updateCandidate(final ServerAddress serverAddress, final ClusterType clusterType) {
+            candidate = isEnabled(clusterType) ? serverAddress : null;
+        }
+
+        public void onAttemptFailure(final Throwable failure) {
+            if (candidate == null || failure instanceof MongoConnectionPoolClearedException) {
+                candidate = null;
+                return;
+            }
+            deprioritized.add(candidate);
+        }
+
+        private static boolean isEnabled(final ClusterType clusterType) {
+            return clusterType == ClusterType.SHARDED;
+        }
+
+        /**
+         * {@link ServerSelector} requires thread safety, but that is only because a user may specify
+         * {@link com.mongodb.connection.ClusterSettings.Builder#serverSelector(ServerSelector)},
+         * which indeed may be used concurrently. {@link DeprioritizingSelector} does not need to be thread-safe.
+         */
+        private final class DeprioritizingSelector implements ServerSelector {
+            private DeprioritizingSelector() {
+            }
+
+            @Override
+            public List<ServerDescription> select(final ClusterDescription clusterDescription) {
+                List<ServerDescription> serverDescriptions = clusterDescription.getServerDescriptions();
+                if (!isEnabled(clusterDescription.getType())) {
+                    return serverDescriptions;
+                } else {
+                    List<ServerDescription> nonDeprioritizedServerDescriptions = serverDescriptions
+                            .stream()
+                            .filter(serverDescription -> !deprioritized.contains(serverDescription.getAddress()))
+                            .collect(toList());
+                    return nonDeprioritizedServerDescriptions.isEmpty() ? serverDescriptions : nonDeprioritizedServerDescriptions;
+                }
```

Comment:
```suggestion
                List<ServerDescription> serverDescriptions = clusterDescription.getServerDescriptions();
                if (!isEnabled(clusterDescription.getType())) {
                    return serverDescriptions;
                }
                List<ServerDescription> nonDeprioritizedServerDescriptions = serverDescriptions
                        .stream()
                        .filter(serverDescription -> !deprioritized.contains(serverDescription.getAddress()))
                        .collect(toList());
                return nonDeprioritizedServerDescriptions.isEmpty() 
                        ? serverDescriptions : nonDeprioritizedServerDescriptions;
```

Reply:
Done in https://github.com/mongodb/mongo-java-driver/pull/1367/commits/246353f36dae51e28aa212488444ec54cb8ad5af.